src/parsing/index.ts

```diff
@@ -66,127 +66,194 @@ type ParsePaymentDestinationArgs = {
+  return inputDataToObject(data)?.query?.lightning
+}
+
+const getProtocolAndData = (
+  destination: string,
+): { protocol: string; destinationText: string } => {
   // input might start with 'lightning:', 'bitcoin:'
   const split = destination.split(":")
   const protocol = split[0].toLocaleLowerCase()
   const destinationText = split[1] ?? split[0]
+  return { protocol, destinationText }
+}
 
+const getDataToDecode = (inputData: string): string => {
+  const lnParam = getLNParam(inputData)
+  if (lnParam != null) {
+    return lnParam
+  }
+  const { protocol, destinationText } = getProtocolAndData(inputData)
+  // some apps encode lightning invoices in UPPERCASE
+  return (
+    protocol.toLowerCase() === "lightning" ? destinationText : protocol
+  ).toLowerCase()
+}
+
+const getPaymentType = ({
+  protocol,
+  destinationText,
+}: {
+  protocol: string
+  destinationText: string
+}): PaymentType => {
   if (destinationText.match(/^lnurl/iu)) {
+    return "lnurl"
+  }
+  if (
+    protocol === "lightning" ||
+    destinationText.match(/^ln(bc|tb).{50,}/iu) ||
+    (destinationText && getLNParam(destinationText) != null)
+  ) {
+    return "lightning"
+  }
+  if (protocol === "onchain" || destinationText.match(/^(1|3|bc1|tb1|bcrt1)/iu)) {
+    return "onchain"
+  }
+  return "intraledger"
+}
+
+const getLNURLPayResponse = ({
+  destinationText,
+}: {
+  destinationText: string
+}): ValidPaymentReponse => {
+  return {
+    valid: true,
+    paymentType: "lnurl",
+    lnurl: destinationText,
+  }
+}
+
+const getLightningPayResponse = ({
+  destination,
+  network,
+  pubKey,
+}: {
+  destination: string
+  network: Network
+  pubKey: string
+}): ValidPaymentReponse => {
+  const paymentType = "lightning"
+  const { destinationText } = getProtocolAndData(destination)
+  const lnProtocol = getLNParam(destination) ?? destinationText
+  if (
+    (network === "mainnet" &&
+      !(lnProtocol.match(/^lnbc/iu) && !lnProtocol.match(/^lnbcrt/iu))) ||
+    (network === "testnet" && !lnProtocol.match(/^lntb/iu)) ||
+    (network === "regtest" && !lnProtocol.match(/^lnbcrt/iu))
+  ) {
     return {
-      valid: true,
-      paymentType: "lnurl",
-      lnurl: destinationText,
+      valid: false,
+      paymentType,
+      paymentRequest: destinationText,
+      errorMessage: `Invalid lightning invoice for ${network} network`,
     }
   }
 
-  if (protocol === "lightning" || destinationText.match(/^ln(bc|tb).{50,}/iu)) {
-    if (
-      (network === "mainnet" && !destinationText.match(/^lnbc/iu)) ||
-      (network === "testnet" && !destinationText.match(/^lntb/iu)) ||
-      (network === "regtest" && !destinationText.match(/^lnbcrt/iu))
-    ) {
-      return {
-        valid: false,
-        paymentType: "lightning",
-        paymentRequest: destinationText,
-        errorMessage: `Invalid lightning invoice for ${network} network`,
-      }
-    }
+  const dataToDecode = getDataToDecode(destinationText)
```

Comment:
Maybe pass protocol to this one too (like intraledger one) and avoid doing this again?

Reply:
I didn't feel comfortable passing in protocol since I needed it for getLNParam, but pointing this out did help me remove this duplicate function we didn't need.